kube/namespace_watch.go

```diff
@@ -0,0 +1,133 @@
+package kube
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/pkg/errors"
+	"github.com/utilitywarehouse/kube-applier/log"
+	"github.com/utilitywarehouse/kube-applier/metrics"
+)
+
+type namespaceWatcher struct {
+	client       kubernetes.Interface
+	resyncPeriod time.Duration
+	stopChannel  chan struct{}
+	store        cache.Store
+	Metrics      metrics.PrometheusInterface
+}
+
+func newNamespaceWatcher(client kubernetes.Interface, resyncPeriod time.Duration, metrics metrics.PrometheusInterface) *namespaceWatcher {
+	return &namespaceWatcher{
+		client:       client,
+		resyncPeriod: resyncPeriod,
+		stopChannel:  make(chan struct{}),
+		Metrics:      metrics,
+	}
+}
+
+func (nw *namespaceWatcher) updateNamespaceMetrics(ns *v1.Namespace) {
+	name := ns.Name
+
+	enabled, err := strconv.ParseBool(ns.Annotations[enabledAnnotation])
+	if err != nil {
+		log.Logger.Error(
+			"Error parsing namespace annotation kube-applier.io/enabled, metric not updated",
+			"namespace", name,
+			"error", err,
+		)
```

Comment:
I think we should be doing something more in the case of missing annotations:
* Delete the metric?
* Or, set it to the default value? This represents the internal value of this setting from kube-applier's perspective.

For instance, if `kube-applier.io/enabled` is removed from a namespace, and its previous value was `"true"`, we will continue to export `kube_applier_enabled 1` which is deceiving because kube-applier won't run for that namespace.

Reply:
good point. will need a bit of redesign so that we do not try to delete stuff that did not exist in the first place. I will try some refactoring.

Reply:
actually it should be really safe to call a delete: https://godoc.org/github.com/prometheus/client_golang/prometheus#GaugeVec.Delete